### PR TITLE
Codomain per channel db

### PR DESCRIPTION
**Pull request by jburel**
# What this PR does

Allow to save the codomain map context per channel instead of per rendering

# Testing this PR
Check that the scripts are correct.
Nobody has ever used the codomain
If somebody did, the rendering engine cannot start due to a ``ClassCastException``
This has been fixed in another PR 

# Related reading

see https://trello.com/c/9iAe3Sg9/95-lut-support

cc @mtbc

### CODE REVIEW

File: components/rendering/src/omeis/providers/re/Renderer.java [hunk 1]
```diff
@@ -704,12 +703,14 @@ public void setCodomainInterval(int start, int end) {
         qd.setCdStart(Integer.valueOf(start));
         qd.setCdEnd(Integer.valueOf(end));
         ome.model.display.CodomainMapContext mapCtx;
+        /*
```
Comment (mtbc) [inline components/rendering/src/omeis/providers/re/Renderer.java]: maybe TODO? perhaps even a card

Comment (jburel) [inline components/rendering/src/omeis/providers/re/Renderer.java]: This is part of work on the card listed in the description.

File: sql/psql/OMERO5.3DEV__10/OMERO5.2__0.sql [hunk 2]
```diff
@@ -1865,16 +1865,25 @@ CREATE INDEX experimenter_config_value ON experimenter_config(value);
...
 
+-- ... up to patch 10:
+
+ALTER TABLE codomainmapcontext 
+    ADD CONSTRAINT FKcodomainmapcontext_channelBinding_channelbinding 
+    FOREIGN KEY (channelBinding) 
+    REFERENCES channelbinding  ;;
+
+ALTER TABLE codomainmapcontext 
+    DROP FOREIGN KEY (renderingDef)
```
Comment (mtbc) [inline sql/psql/OMERO5.3DEV__10/OMERO5.2__0.sql]:
```
ERROR:  syntax error at or near "FOREIGN"
```

You probably need to drop the constraint.

File: sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU General Public License for more details.
+--
+-- You should have received a copy of the GNU General Public License along
+-- with this program; if not, write to the Free Software Foundation, Inc.,
+-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+--
+
+---
+--- OMERO5 development release upgrade from OMERO5.3DEV__8 to OMERO5.3DEV__10.
```
Comment (mtbc) [inline sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql]: from 9

**Comment by mtbc** [comment]
The upgrade script needs also to address the changed uniqueness constraint on the `codomainmapcontext` table.

**Comment by mtbc** [comment]
(I wonder why `schema.sql` got such a horrifying diff, I think the changes are probably actually fine!)

**Comment by jburel** [comment]
@mtbc: thanks for spotting the uniqueness point.

**Comment by jburel** [comment]
the ``schema.sql`` was generated using the ``build-schema`` target

**Comment by mtbc** [comment]
I expect it's fine: I wonder if there's something we could tweak in our code generation to, say, make ordering more predictable, unless we're just catching an edge case in the diff algorithm.

**Comment by mtbc** [comment]
Good to merge.

**Comment by jburel** [comment]
removing breaking label for tomorrow

**Comment by jburel** [comment]
Everything is back to normal
https://ci.openmicroscopy.org/view/OMERO/job/OMERO-DEV-merge-integration-Python27/300/

**Comment by pwalczysko** [comment]
The problems from yesterday are gone, insight is working normally during testing of https://github.com/openmicroscopy/openmicroscopy/pull/4796

File: sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+      -- At most, there should be one duplicate
+      SELECT id INTO duplicate
+        FROM codomainmapcontext
+       WHERE channelBinding = new.channelBinding AND channelBinding_index = new.channelBinding_index
+      OFFSET 0
+       LIMIT 1;
+
+      IF duplicate IS NOT NULL THEN
+          RAISE NOTICE ''Remapping codomainmapcontext %% via (-1 - oldvalue )'', duplicate;
```
Comment (mtbc) [inline sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql]: Bah, missed this: it should probably be a single `%` in this script?

Comment (jburel) [inline sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql]: Do you want to open a followup PR?

Comment (mtbc) [inline sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql]: I will, it looks like there are more to fix too, reported by @chris-allan.

Comment (mtbc) [inline sql/psql/OMERO5.3DEV__10/OMERO5.3DEV__9.sql]: Fixed in #4822.

**Pull request merged**